src/components/adaptive-cards/AdaptiveCard/AdaptiveCard.jsx

```diff
@@ -96,21 +100,56 @@ export default function AdaptiveCard({
       };
     });
   };
+
   const setInput = useCallback((input) => {
     setInputs((prevInputs) => ({...prevInputs, [input.id]: input}));
   }, [setInputs]);
-  const getValue = (id, defval = '') => ((id in inputs && inputs[id].value) ? inputs[id].value : defval);
-  const getAllValues = () => Object.entries(inputs).reduce((allValues, [id, input]) => (
-    {...allValues, [id]: input.value}
-  ), {});
+
+  const getValue = (id, defval = '') => ((id in inputs && inputs[id].value !== undefined) ? inputs[id].value : defval);
+
+  const getAllValues = () => Object.fromEntries(Object.entries(inputs).map(
+    ([id, input]) => [id, input.value],
+  ));
+
+  const getError = (id) => inputs[id]?.error;
+
+  const validate = () => {
+    let valid = true;
+
+    Object.values(inputs).forEach((input) => {
+      let error;
+
+      if (input.isRequired && !input.value && input.value !== 0) {
+        error = input.errorMessage;
+      } else if (input.value < input.min) {
+        error = `Minimum value is ${input.min}`;
+      } else if (input.value > input.max) {
+        error = `Maximum value is ${input.max}`;
+      } else if (input.regex && !String(input.value).match(input.regex)) {
+        error = input.errorMessage;
+      } else if (String(input.value).length > input.maxLength) {
+        error = `Maximum length is ${input.maxLength}`;
+      }
+
+      valid = valid && !error;
+      setInput({...input, error});
```

Comment:
to minimize the number of `setInput` calls (which are all changing state), let's collect all inputs and do a final `setInputs` instead:
```js
let newInputs = Object.fromEntries(Object.entries(inputs).map([id, input] => {
    ...
    return [id, {...input, error}];
}));

setInputs(newInputs);

return Object.values(newInputs).some(input => input.error);

Reply:
and it seems we are mapping object values more than once (at least in this PR), so let's create a helper function for mapping values of an object, and put it in util.js:

```js
export function mapValues(object, mapper) {
  return Object.fromEntries(Object.entries(object).map([key, value] => (
    [key, mapper(value, key)]
  )));
}
```
and use it like this:

```js
const newInputs = mapValues(inputs, (input) => {
  ...
  return {...input, error};
});